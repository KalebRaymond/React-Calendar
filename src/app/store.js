import { configureStore } from '@reduxjs/toolkit'
import calendarReducer from '../reducers/calendarReducer.js'

export default configureStore({
  reducer: {
    calendar: calendarReducer
  }
})