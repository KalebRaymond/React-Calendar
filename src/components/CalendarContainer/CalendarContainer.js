import { fetchEvents } from "../../reducers/calendarReducer";
import { useDispatch, useSelector } from "react-redux";
import { useTranslation } from "react-i18next";
import CalendarGrid from "components/CalendarGrid/CalendarGrid";
import CalendarGridCard from "components/CalendarGridCard/CalendarGridCard";
import CalendarGridRow from "components/CalendarGridRow/CalendarGridRow";
import moment from "moment";
import React, { useContext, useEffect } from "react";
import styles from "./CalendarContainer.scss";
import TranslationService from "services/TranslationService";
import { ThemeContext } from "context/ThemeContext";

///What is the point of this component? Delete it?
const CalendarContainer = () => {
	const { t } = useTranslation();
	const dispatch = useDispatch();
	const { theme } = useContext(ThemeContext);

	///Make these two hooks into a custom hook?
	//Keep track of first and last visible dates when the focused date changes
	const [firstVisibleDate, lastVisibleDate] = useSelector((state) => {
		return [
			state.calendar.visibleDates[0][0],
			state.calendar.visibleDates[state.calendar.visibleDates.length - 1][6],
		];
	});

	//Fetch events when the first and last visible dates change
	useEffect(() => {
		if (firstVisibleDate && lastVisibleDate) {
			dispatch(fetchEvents(firstVisibleDate, lastVisibleDate));
		}
	}, [firstVisibleDate, lastVisibleDate]);

	//Index of the month that is currently focused
	const focusedMonth = useSelector((state) =>
		moment(state.calendar.focusedDate).month()
	);
	//Year that is currently focused
	const focusedYear = useSelector((state) =>
		moment(state.calendar.focusedDate).year()
	);
	//Todays date, needed to determine if a date is today's date and display it differently
	const todaysDate = moment();

	/*	Given an object where each key maps a date to an array of objects
	 *	representing the events that start on that date, returns a nested array
	 *	that represents how the events should be displayed on the calendar.
	 *
	 * 	Returns an array containing an array for each calendar row, which contains
	 * 	objects that have a "multiDateEvents" array and a "singleDateEvents" array
	 */
	const generateEventMatrix = (events, visibleDates) => {
		//Create a matrix of empty arrays for each day of the month
		const eventMatrix = visibleDates.map((week) => {
			return week.map((_) => {
				return {
					multiDateEvents: [],
					singleDateEvents: [],
				};
			});
		});

		const curEventPositions = [];

		//Add events to the matrix in the correct position that they will
		//be displayed in the calendar. In order to make the multidate events
		//get drawn in the correct vertical position, need to put a copy of the
		//event in every day of the event's duration in the correct vertical position.
		for (let week = 0; week < eventMatrix.length; week++) {
			for (let day = 0; day < eventMatrix[week].length; day++) {
				//Get the current date string
				///Replace visibleDates with a moment().format() string TODO
				const curDate = moment()
					.set({
						date: visibleDates[week][day].date,
						month: visibleDates[week][day].month,
						year: visibleDates[week][day].year,
					})
					.format("YYYY-MM-DD");

				//Get events that start on this day
				//evens[curDate] contains all events that occur on this date,
				//including multi-day events that may start on an earlier date
				const unseenEvents = events[curDate]
					? events[curDate].filter((event) => {
							return event.startDate === curDate;
					  })
					: [];

				//Add events into first availble slot in curEventPositions
				for (let i = 0; i < unseenEvents.length; i++) {
					insertIntoFirstNullIndex(curEventPositions, unseenEvents[i]);
				}

				//Add events to the matrix
				for (let i = 0; i < curEventPositions.length; i++) {
					if (curEventPositions[i]) {
						///Ugly hack to make a deep copy of the event object
						const event = JSON.parse(JSON.stringify(curEventPositions[i]));

						//If the event is a single date event, add it to the singleDateEvents array
						if (event.numDays == 1) {
							eventMatrix[week][day].singleDateEvents.push(event);
						}
						//If the event is a multi date event, add it to the multiDateEvents array
						else {
							//If the event starts on this day or curDate is the first
							//day of the week, set isStartOfButton to true
							if (event.startDate === curDate || day === 0) {
								event.isStartOfButton = true;

								//Set buttonLength to the number of days the event lasts
								//or the number of days left in the week, whichever is smaller
								const daysLeftInWeek = 7 - day;
								const daysLeftInEvent =
									moment(event.endDate).diff(moment(curDate), "days") + 1;

								event.buttonLength = Math.min(daysLeftInWeek, daysLeftInEvent);
							}
							//If the event does not start on this day, set isStartOfButton to false
							else {
								event.isStartOfButton = false;
							}

							eventMatrix[week][day].multiDateEvents.push(event);
						}
					}
				}

				//Remove events that end on this day from curEventPositions
				///Might need to move this block down?
				for (let i = 0; i < curEventPositions.length; i++) {
					if (
						curEventPositions[i] &&
						curEventPositions[i].endDate === curDate
					) {
						curEventPositions[i] = null;
					}
				}
			}
		}

		return eventMatrix;
	};

	//Inserts a value into the first null index of an array,
	//or at the end if no indices have a null value.
	//Helper function for generateEventMatrix
	const insertIntoFirstNullIndex = (arr, val) => {
		for (let i = 0; i < arr.length; i++) {
			if (arr[i] === null) {
				arr[i] = val;
				return;
			}
		}

		arr.push(val);
	};

	//2D array containing the events that should be displayed on each visible date
	const eventMatrix = useSelector((state) => {
		return generateEventMatrix(
			state.calendar.events,
			state.calendar.visibleDates
		);
	});

	///Shallow copy of dates object - performance hit?
	const renderCalendarContent = ([...dates]) => {
		let calendarContent = [];

		calendarContent = dates.map((row, i) => (
			<CalendarGridRow key={`row-${focusedMonth}-${focusedYear}-${i}`}>
				{row.map((dateObject, j) => (
					<CalendarGridCard
						cardAriaLabel={`${TranslationService.getWeekdayFromIndex(j)}, ${t(
							"dateFormats.MDY",
							{
								month: TranslationService.getMonthTranslation(dateObject.month),
								day: TranslationService.getOrdinal(dateObject.date),
								year: dateObject.year,
							}
						)}`}
						date={moment().set({
							date: dateObject.date,
							month: dateObject.month,
							year: dateObject.year,
						})}
						events={eventMatrix[i][j]}
						grayed={dateObject.month !== focusedMonth}
						isTodaysDate={
							dateObject.month === todaysDate.month() &&
							dateObject.year === todaysDate.year() &&
							dateObject.date === todaysDate.date()
						}
						key={`date-${dateObject.month}-${dateObject.date}-${dateObject.year}`}
					></CalendarGridCard>
				))}
			</CalendarGridRow>
		));

		return calendarContent;
	};

	//Render the calendar content when the state changes
	const calendarContent = useSelector((state) =>
		renderCalendarContent(state.calendar.visibleDates)
	);

	return (
		<div
			className={`CalendarContainer ${theme === "light" ? "light" : "dark"}`}
			data-testid="CalendarContainer"
		>
			<CalendarGrid>{calendarContent}</CalendarGrid>
		</div>
	);
};

export default CalendarContainer;
