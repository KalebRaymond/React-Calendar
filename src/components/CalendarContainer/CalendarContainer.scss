.CalendarContainer {
	height: 100%;
	width: 100%;

	&.light {
		background-color: #fff;
	}

	&.dark {
		background-color: #333;
		color: #fff;
	}
}
