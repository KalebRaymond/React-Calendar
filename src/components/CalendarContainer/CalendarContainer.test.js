import "@testing-library/jest-dom/extend-expect";
import { Provider } from "react-redux";
import { render, screen } from "@testing-library/react";
import CalendarContainer from "./CalendarContainer";
import React from "react";
import store from "../../app/store.js";
import i18n from "i18n"; //Required because CalendarContainer uses TranslationService

describe("<CalendarContainer />", () => {
	test("it should mount", () => {
		render(
			<Provider store={store}>
				<CalendarContainer />
			</Provider>
		);

		const calendarContainer = screen.getByTestId("CalendarContainer");

		expect(calendarContainer).toBeInTheDocument();
	});
});
