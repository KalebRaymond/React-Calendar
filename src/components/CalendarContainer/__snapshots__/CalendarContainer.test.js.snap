// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalendarContainer /> should render events correctly 1`] = `
<div
  class="CalendarContainer light"
  data-testid="CalendarContainer"
>
  <div
    aria-label="Calendar Grid"
    class="CalendarGrid"
    data-testid="CalendarGrid"
    role="grid"
  >
    <div
      aria-label="Week Day Header"
      class="WeekDaysHeader"
      data-testid="WeekDaysHeader"
      role="row"
    >
      <div
        class="weekDay"
        role="columnheader"
        tabindex="-1"
      >
        <span>
          SUN
        </span>
      </div>
      <div
        class="weekDay"
        role="columnheader"
        tabindex="-1"
      >
        <span>
          MON
        </span>
      </div>
      <div
        class="weekDay"
        role="columnheader"
        tabindex="-1"
      >
        <span>
          TUE
        </span>
      </div>
      <div
        class="weekDay"
        role="columnheader"
        tabindex="-1"
      >
        <span>
          WED
        </span>
      </div>
      <div
        class="weekDay"
        role="columnheader"
        tabindex="-1"
      >
        <span>
          THU
        </span>
      </div>
      <div
        class="weekDay"
        role="columnheader"
        tabindex="-1"
      >
        <span>
          FRI
        </span>
      </div>
      <div
        class="weekDay"
        role="columnheader"
        tabindex="-1"
      >
        <span>
          SAT
        </span>
      </div>
    </div>
    <div
      class="CalendarGridRow"
      data-testid="CalendarGridRow"
      role="row"
    >
      <div
        class="CalendarGridCard grayed light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            26
          </span>
        </div>
        <button
          aria-label="Sunday, February 26th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard grayed light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            27
          </span>
        </div>
        <button
          aria-label="Monday, February 27th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard grayed light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            28
          </span>
        </div>
        <button
          aria-label="Tuesday, February 28th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            1
          </span>
        </div>
        <button
          aria-label="Wednesday, March 1st, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            2
          </span>
        </div>
        <button
          aria-label="Thursday, March 2nd, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            3
          </span>
        </div>
        <button
          aria-label="Friday, March 3rd, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            4
          </span>
        </div>
        <button
          aria-label="Saturday, March 4th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
    </div>
    <div
      class="CalendarGridRow"
      data-testid="CalendarGridRow"
      role="row"
    >
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            5
          </span>
        </div>
        <button
          aria-label="Sunday, March 5th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            6
          </span>
        </div>
        <button
          aria-label="Monday, March 6th, 2023  1 event on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label=""
          class="eventListContainer"
          role="region"
        >
          <div
            class="EventListContainer"
            data-testid="EventListContainer"
          >
            <ul
              aria-label="Event List"
              class="eventList"
            >
              <li>
                <button
                  aria-label="Test, event from 2023-03-06 to 2023-03-08"
                  class="eventButton multiDateEventButton"
                  data-testid="MultipleDateEvent"
                  style="width: calc((100% * 3  - 0.15rem);"
                >
                  <span
                    class="eventName"
                  >
                    Test
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            7
          </span>
        </div>
        <button
          aria-label="Tuesday, March 7th, 2023  2 events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label=""
          class="eventListContainer"
          role="region"
        >
          <div
            class="EventListContainer"
            data-testid="EventListContainer"
          >
            <ul
              aria-label="Event List"
              class="eventList"
            >
              <li>
                <button
                  aria-label="Test, event from 2023-03-06 to 2023-03-08"
                  class="eventButton placeholderEventButton"
                  data-testid="MultipleDateEvent"
                />
              </li>
              <li>
                <button
                  aria-label="Huh, event from 2023-03-07 to 2023-03-09"
                  class="eventButton multiDateEventButton"
                  data-testid="MultipleDateEvent"
                  style="width: calc((100% * 3  - 0.15rem);"
                >
                  <span
                    class="eventName"
                  >
                    Huh
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            8
          </span>
        </div>
        <button
          aria-label="Wednesday, March 8th, 2023  2 events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label=""
          class="eventListContainer"
          role="region"
        >
          <div
            class="EventListContainer"
            data-testid="EventListContainer"
          >
            <ul
              aria-label="Event List"
              class="eventList"
            >
              <li>
                <button
                  aria-label="Test, event from 2023-03-06 to 2023-03-08"
                  class="eventButton placeholderEventButton"
                  data-testid="MultipleDateEvent"
                />
              </li>
              <li>
                <button
                  aria-label="Huh, event from 2023-03-07 to 2023-03-09"
                  class="eventButton placeholderEventButton"
                  data-testid="MultipleDateEvent"
                />
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            9
          </span>
        </div>
        <button
          aria-label="Thursday, March 9th, 2023  1 event on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label=""
          class="eventListContainer"
          role="region"
        >
          <div
            class="EventListContainer"
            data-testid="EventListContainer"
          >
            <ul
              aria-label="Event List"
              class="eventList"
            >
              <li>
                <button
                  aria-label="Huh, event from 2023-03-07 to 2023-03-09"
                  class="eventButton placeholderEventButton"
                  data-testid="MultipleDateEvent"
                />
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            10
          </span>
        </div>
        <button
          aria-label="Friday, March 10th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            11
          </span>
        </div>
        <button
          aria-label="Saturday, March 11th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
    </div>
    <div
      class="CalendarGridRow"
      data-testid="CalendarGridRow"
      role="row"
    >
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            12
          </span>
        </div>
        <button
          aria-label="Sunday, March 12th, 2023  1 event on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label=""
          class="eventListContainer"
          role="region"
        >
          <div
            class="EventListContainer"
            data-testid="EventListContainer"
          >
            <ul
              aria-label="Event List"
              class="eventList"
            >
              <li>
                <button
                  aria-label="Test, event starting at 10:00am on 2023-03-12"
                  class="SingleDateEvent light eventButton"
                  data-testid="SingleDateEvent"
                >
                  <div
                    class="bullet"
                  />
                  <span
                    class="eventDetails"
                  >
                    <span
                      class="eventTime"
                    >
                      10:00am
                    </span>
                    <strong
                      class="eventName"
                    >
                      Test
                    </strong>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            13
          </span>
        </div>
        <button
          aria-label="Monday, March 13th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            14
          </span>
        </div>
        <button
          aria-label="Tuesday, March 14th, 2023  2 events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label=""
          class="eventListContainer"
          role="region"
        >
          <div
            class="EventListContainer"
            data-testid="EventListContainer"
          >
            <ul
              aria-label="Event List"
              class="eventList"
            >
              <li>
                <button
                  aria-label="Test, event starting at 10:00am on 2023-03-14"
                  class="SingleDateEvent light eventButton"
                  data-testid="SingleDateEvent"
                >
                  <div
                    class="bullet"
                  />
                  <span
                    class="eventDetails"
                  >
                    <span
                      class="eventTime"
                    >
                      10:00am
                    </span>
                    <strong
                      class="eventName"
                    >
                      Test
                    </strong>
                  </span>
                </button>
              </li>
              <li>
                <button
                  aria-label="Test, event starting at 12:00pm on 2023-03-14"
                  class="SingleDateEvent light eventButton"
                  data-testid="SingleDateEvent"
                >
                  <div
                    class="bullet"
                  />
                  <span
                    class="eventDetails"
                  >
                    <span
                      class="eventTime"
                    >
                      12:00pm
                    </span>
                    <strong
                      class="eventName"
                    >
                      Test
                    </strong>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            15
          </span>
        </div>
        <button
          aria-label="Wednesday, March 15th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            16
          </span>
        </div>
        <button
          aria-label="Thursday, March 16th, 2023  1 event on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label=""
          class="eventListContainer"
          role="region"
        >
          <div
            class="EventListContainer"
            data-testid="EventListContainer"
          >
            <ul
              aria-label="Event List"
              class="eventList"
            >
              <li>
                <button
                  aria-label="Midnight, event starting at 12:00pm on 2023-03-16"
                  class="SingleDateEvent light eventButton"
                  data-testid="SingleDateEvent"
                >
                  <div
                    class="bullet"
                  />
                  <span
                    class="eventDetails"
                  >
                    <span
                      class="eventTime"
                    >
                      12:00pm
                    </span>
                    <strong
                      class="eventName"
                    >
                      Midnight
                    </strong>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            17
          </span>
        </div>
        <button
          aria-label="Friday, March 17th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            18
          </span>
        </div>
        <button
          aria-label="Saturday, March 18th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
    </div>
    <div
      class="CalendarGridRow"
      data-testid="CalendarGridRow"
      role="row"
    >
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            19
          </span>
        </div>
        <button
          aria-label="Sunday, March 19th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date todaysDate"
            data-testid="card-date"
          >
            20
          </span>
        </div>
        <button
          aria-label="Monday, March 20th, 2023 Today's date No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            21
          </span>
        </div>
        <button
          aria-label="Tuesday, March 21st, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            22
          </span>
        </div>
        <button
          aria-label="Wednesday, March 22nd, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            23
          </span>
        </div>
        <button
          aria-label="Thursday, March 23rd, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            24
          </span>
        </div>
        <button
          aria-label="Friday, March 24th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            25
          </span>
        </div>
        <button
          aria-label="Saturday, March 25th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
    </div>
    <div
      class="CalendarGridRow"
      data-testid="CalendarGridRow"
      role="row"
    >
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            26
          </span>
        </div>
        <button
          aria-label="Sunday, March 26th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            27
          </span>
        </div>
        <button
          aria-label="Monday, March 27th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            28
          </span>
        </div>
        <button
          aria-label="Tuesday, March 28th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            29
          </span>
        </div>
        <button
          aria-label="Wednesday, March 29th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            30
          </span>
        </div>
        <button
          aria-label="Thursday, March 30th, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard  light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            31
          </span>
        </div>
        <button
          aria-label="Friday, March 31st, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
      <div
        class="CalendarGridCard grayed light"
        data-testid="CalendarGridCard"
        role="gridcell"
      >
        <div
          class="dateContainer"
        >
          <span
            class="date "
            data-testid="card-date"
          >
            1
          </span>
        </div>
        <button
          aria-label="Saturday, April 1st, 2023  No events on this date  Press ENTER to add a new event on this date"
          class="cardButton"
          data-testid="cardButton"
        />
        <div
          aria-label="No events on this date"
          class="eventListContainer"
          role="region"
        />
      </div>
    </div>
  </div>
</div>
`;
